### PR TITLE
Add binary reuse to Verilator tests via annotation

### PR DESCRIPTION
**Pull request by ryan-lund**
Adds a CachingAnnotation which enables binary reuse across verilator test runs if high FIRRTL, elaborated annotations, and underlying versions match. In testing reduces the VerilatorBasicTests runtime from 35 seconds on first run to 5 seconds on second (cached) run.

Addresses: https://github.com/ucb-bar/chisel-testers2/issues/106 and https://github.com/ucb-bar/chisel-testers2/issues/212

### CODE REVIEW

**Review by ducky64 — COMMENTED** [review]
This looks fine at a high level, though I'd prefer it be explicitly in the experimental or advanced user package, because tests now depend on state on disk. This also differs from the proposal in the linked issues, which would provide users with an explicit reusable handle to a compiled simulator (essentially separating the compilation and simulation stages), instead of implicitly re-using cached binaries.

For implementation: unless this is only for super advanced users running a *nix-like OS, consider seeing if there is a Java/Scala API for file operations, so it's more platform-neutral.

I'd also prefer that there's no printing, since unit tests should not require users to check the console output (and console spam also isn't helpful for regressions),

**Comment by ducky64** [comment]
Also, consider adding documentation on the mechanism of action, such as in the scaladoc for the annotation.

File: build.sbt [hunk 1]
```diff
@@ -16,7 +16,9 @@ resolvers ++= Seq(
 
 libraryDependencies ++= Seq(
   "org.scalatest" %% "scalatest" % "3.2.2",
-  "com.lihaoyi" %% "utest" % "latest.integration"
+  "com.lihaoyi" %% "utest" % "latest.integration",
+  "com.github.plokhotnyuk.jsoniter-scala" %% "jsoniter-scala-core"   % "2.6.4",
+  "com.github.plokhotnyuk.jsoniter-scala" %% "jsoniter-scala-macros" % "2.6.4" % "provided"
```
Comment (ekiwi) [inline build.sbt:21]: I don't think we should be adding a second json library to our dependencies.
Could you check to see how hard it would be to use [json4s](https://github.com/json4s/json4s) instead?
That is the library used by `firrtl` and thus already pulled into `chiseltest`.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 2]
```diff
@@ -48,6 +52,74 @@ object VerilatorExecutive extends BackendExecutive {
...
+      val highFirrtlPath = s"${targetDir}/${circuit.name}.hi.fir"
+      val oldHighFirrtlPath = s"${targetDir}/${circuit.name}.hi.fir.old"
+      s"mv ${highFirrtlPath} ${oldHighFirrtlPath}".!
+
+      (new ChiselStage).run(
+        elaboratedAnno :+ RunFirrtlTransformAnnotation(new HighFirrtlEmitter)
+      )
+
+      // Diff high FIRRTL and delete old file
+      val diffExit = s"diff ${highFirrtlPath} ${oldHighFirrtlPath}".!
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:66]: Instead of relying on GNU tools here, you could try to use the hashing mechanism that we already have for firrtl:
https://github.com/chipsalliance/firrtl/blob/edb91f7bc613026f824519786c3ce25740bb21c3/src/main/scala/firrtl/ir/StructuralHash.scala#L29

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 3]
```diff
@@ -48,6 +52,74 @@ object VerilatorExecutive extends BackendExecutive {
...
+
+      // Hash the deterministic elements of elaboratedAnno for comparison
+      // currently since this is guaranteed to be in Verilator and have the same test run dir, this should
+      // only catch if coverage or vcd flags change
+      val elaboratedAnnoHash = elaboratedAnno
+        .toSeq
+        .filter(anno => !List("chisel3.stage.ChiselCircuitAnnotation", "chisel3.stage.DesignAnnotation")
+          .contains(anno.getClass.toString.split(" ").last))
+        .map(anno => anno.hashCode)
+      val firrtlInfoHash = Seq(firrtl.BuildInfo.hashCode)
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:80]: This isn't really good enough since `hashCode` is not a cryptographic hash, it is only really meant to quickly discover that two things are _not_ equal, but it isn't good enough for equality.
The easiest approach here might be to serialize all annotations to a JSON string and then use a Java `MessageDigest` to calculate a sha256 hash.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 4]
```diff
@@ -48,6 +52,74 @@ object VerilatorExecutive extends BackendExecutive {
...
+      val systemVersionHash = Seq(System.getProperty("java.version"), scala.util.Properties.versionNumberString)
+        .map(version => version.hashCode)
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:82]: again, `hashCode` is wrong

**Review by ekiwi — CHANGES_REQUESTED** [review]
I think the general idea is good. The two things that I would like to ask you to change are:
1. use json4s instead of jsoniter
2. rethink how you do the hash, see my comments

**Comment by ekiwi** [comment]
@jackkoenig:  We could add the build time to the firrtl build info: https://github.com/sbt/sbt-buildinfo

**Comment by oharboe** [comment]
This would speed up our tests tremendously. 

Note that when running tests in parallel, Verilator and Chisel elaboration are the memory hogs. 

There should be a way to run a pre-processing stage that does Verilator and Chisel elaboration in parallel to the extent that there's memory available. Normally with Chisel I run out of memory before cores when running things in parallel. Especially in the cloud where there's ca. 1gByte per core for e.g. Google Cloud Build machine profiles.

File: build.sbt (hunk 1 above)
Comment (ryan-lund) [reply build.sbt:21]: For now I've kept jsoniter because I've found its feature set is better set-up for serializing generic objects without having to write extra converters. (I.E as it is I can serialize the CombinationPath object without the need to write anything special). I haven't been able to replicate that with json4s but I'll keep investigating.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala (hunk 2 above)
Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala]: Changed to use your structural hash (thank you!) and replaced Unix operations in general with native Scala / Java file operations.

**Comment by ryan-lund** [comment]
The largest update I've made in the latest push is a change to the way in which the cache is stored. The goals of these changes are to allow tests with different names or classes but identical DUTs to match the same cache entry. To achieve this, a new directory called cache_dir is created. In this director, the generated Verilog and Verilator sources are stored in a directory named test_run_dir/cache_dir/<high firrtl hash>/<annotations hash>. Thus any DUTs that has the same high firrtl hash and the same annotations will load the same sources. This source location is then simlinked to the specific test directory so that a user can view the sources used for testing (i.e. test_run_dir/should_<test name>/sources -> est_run_dir/cache_dir/<high firrtl hash>/<annotations hash>). Waveform files are always written to the specific test directory so that they are not overwritten by the results of un-related tests.

As before, these features are only turned on when the CachingAnnotation is supplied. I've tested the VerilatorBasicTests without the annotation and verified that the sources generate in the typical  test_run_dir/should_<test name> location.

I will also write up detailed documentation as per Ducky's comment about scaladocs.

**Comment by ekiwi** [comment]
> In this director, the generated Verilog and Verilator sources are stored in a directory named test_run_dir/cache_dir//.

Quick question: How are you making sure that there are no race conditions between two different threads that both try to compile the same DUT?

File: build.sbt (hunk 1 above)
Comment (ekiwi) [reply build.sbt:21]: I don't want to sign off on adding a new dependency. Removing that dependency later will be a great pain, since users might start depending on it. If @ducky64  is OK, with adding it, then we can. However, if it up to just me, then I believe that we should not add this dependency.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala [hunk 5]
```diff
@@ -190,8 +190,8 @@ int main(int argc, char **argv, char **env) {
 #endif
 #if VM_COVERAGE
     VL_PRINTF(\"Writing Coverage..\");
-    Verilated::mkdir("$targetDir/logs");
-    VerilatedCov::write("$targetDir/logs/coverage.dat");
+    Verilated::mkdir("$coverageDir");
+    VerilatedCov::write("$coverageDir/coverage.dat");
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala:194]: How is this change related to this PR? Did you add this by accident?

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 6]
```diff
@@ -38,16 +45,101 @@ object VerilatorExecutive extends BackendExecutive {
...
+      val highFirrtlAnnos = (new ChiselStage).run(
+        elaboratedAnno :+ RunFirrtlTransformAnnotation(new HighFirrtlEmitter)
+      )
+
+      val rawFirrtl = highFirrtlAnnos
+        .collect { case EmittedFirrtlCircuitAnnotation(e) => e.value}
+        .map(l => firrtl.Parser.parse(l))
+        .head
+
+      val moduleHashString = DatatypeConverter
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:72]: please split this up with some more named `val`s

File: build.sbt (hunk 1 above)
Comment (ducky64) [reply build.sbt:21]: I generally would prefer to not add dependencies, and additionally prefer to standardize libraries across the chisel projects where possible. But if there's some strong rationale for a new library (or if we want to standardize on a new library project-wide, perhaps without necessarily migrating old code), it's worth discussing.

Comment (jackkoenig) [reply build.sbt:21]: We're fairly close to being able to build Chisel and FIRRTL without any runtime reflection except for generic serialization and deserialization of annotations. This isn't like, an important goal, but it would be cool to be able to emit ScalaJS as well as create GraalVM Native images. jsoniter appears to have better support for non-JVM targets than json4s, so it's worth considering using it throughout instead. This is a high code migration cost though since some users have written custom [de]serializers.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala (hunk 6 above)
Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala]: Done!

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala (hunk 5 above)
Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala:194]: This was an intentional change. I realized when making the more global prototype that the coverage directory would always default to $targetDir/logs. I felt that it would be a more useful API to let to caller of the C++ harness generator fully specify the path to a coverage directory.

Comment (ekiwi) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala:194]: I believe this should go into a separate PR

Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorCppHarnessGenerator.scala]: Removed

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 7]
```diff
@@ -48,6 +55,115 @@ object VerilatorExecutive extends BackendExecutive {
...
+      }
+
+      val moduleWriter = new FileWriter(moduleHashFile)
+      moduleWriter.write(moduleHashHex)
+      moduleWriter.close()
+
+      // Hash the deterministic elements of elaboratedAnno for comparison, this should
+      // only catch if coverage or vcd flags change
+      val elaboratedAnnoString = elaboratedAnno.toSeq
+        .filter(anno =>
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:101]: could you replace this with something like?:

```.scala
.filter {
  case _ : chisel3.stage.ChiselCircuitAnnotatio | _ :  chisel3.stage.DesignAnnotation | _ : firrtl.options.TargetDirAnnotation => false
  case _ => true
}
```

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala [hunk 8]
```diff
@@ -48,6 +55,115 @@ object VerilatorExecutive extends BackendExecutive {
     val circuit = elaboratedAnno.collect { case x: ChiselCircuitAnnotation => x }.head.circuit
     val dut = getTopModule(circuit).asInstanceOf[T]
 
+    if (elaboratedAnno.contains(CachingAnnotation)) {
```
Comment (ekiwi) [inline src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:58]: Since the whole thing has gotten quite long, could you please move this into its own function?

File: src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      c.io.out.expect(0.U)
+
+      c.reset.poke(false.B)
+      c.io.in.poke(43.U)
+      c.clock.step()
+      c.io.out.expect(43.U)
+    }
+  }
+
+  it should "test waveform generation" in {
```
Comment (ekiwi) [inline src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala:122]: This should do two checks:

1. check that the VCD is not there before `test`
2. check that the VCD now exists after `test`

File: src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+import chisel3._
+import chiseltest.tests.{PassthroughModule, StaticModule}
+import chiseltest._
+import chiseltest.experimental.TestOptionBuilder._
+import chiseltest.internal.{CachingAnnotation, VerilatorBackendAnnotation, WriteVcdAnnotation}
+import org.scalatest._
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class VerilatorCachingTests extends AnyFlatSpec with ChiselScalatestTester with Matchers {
```
Comment (ekiwi) [inline src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala:13]: Please add a test with multiple calls to `test` inside a single scalatest unit.

**Review by ekiwi — CHANGES_REQUESTED** [review]

File: src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala (hunk 10 above)
Comment (ryan-lund) [reply src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala:16]: Done, added multiple calls with the same module in one case and different modules in another case.

File: src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala (hunk 9 above)
Comment (ryan-lund) [reply src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala]: Updated this test to match the VCD test in OptionsPassingTest.scala

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala (hunk 8 above)
Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala:47]: Moved this into `getCachedSim`. Moved caching generation into `cacheSim`.

File: src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala (hunk 7 above)
Comment (ryan-lund) [reply src/main/scala/chiseltest/legacy/backends/verilator/VerilatorExecutive.scala]: Done

**Review by ekiwi — APPROVED** [review]
Thanks @ryan-lund !

Let's get this in :shipit:

**Comment by oharboe** [comment]
@ryan-lund Thanks! I'm looking forward to testing this! :-)

**Comment by ekiwi** [comment]
> @ryan-lund Thanks! I'm looking forward to testing this! :-)

Yeah, please do let us know if you find any issues with this. That is why we are publishing the `SNAPSHOT` releases, so that the community can test things early and identify any issues that we missed in our testing.

**Comment by oharboe** [comment]
@ekiwi ETA for SNAPSHOT? Is there any doc to go with this or do I just run a test and time it?

**Comment by ekiwi** [comment]
>  ETA for SNAPSHOT? Is there any doc to go with this or do I just run a test and time it?

It is already out: https://github.com/ucb-bar/chisel-testers2/runs/2634441405
Just make sure to clear your coursier cache. sbt might only be checking for new releases every 24h or so.

You can have a look at Ryan's tests to see how to make use of the new feature:
https://github.com/ryan-lund/chisel-testers2/blob/master/src/test/scala/chiseltest/experimental/tests/VerilatorCachingTests.scala

**Comment by oharboe** [comment]
I'm running into build problems when I upgrade...
```
Symbol 'type org.scalatest.FlatSpec' is missing from the classpath.\nThis symbol is required by 'class chisel3.iotesters.ChiselFlatSpec'.\nMake sure that type FlatSpec is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.\nA full rebuild may help if 'ChiselFlatSpec.class' was compiled against an incompatible version of org.scalatest.
```

```
val defaultVersions = Map(
  "chisel3" -> "3.5-SNAPSHOT",
  "chisel-iotesters" -> "1.5-SNAPSHOT",
  "chiseltest" -> "0.5-SNAPSHOT"
)

```

**Comment by oharboe** [comment]
Deleted some tests and now I can build, but I get this error message:

```
[error] java.lang.NoClassDefFoundError: com/vladsch/flexmark/ast/Node
[error]         at org.scalatest.tools.ReporterFactory.createHtmlReporter(ReporterFactory.scala:194)
[error]         at org.scalatest.tools.ReporterFactory.getReporterFromConfiguration(ReporterFactory.scala:245)
[error]         at org.scalatest.tools.ReporterFactory.$anonfun$createReportersFromConfigurations$1(ReporterFactory.scala:254)
[error]         at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:273)
```

**Comment by oharboe** [comment]
I finally ran into a problem that appears to be related to this new feature:

```
 error: redefinition of ‘void chisel_insertp(const char*, const char*, const char*, const char*, const char*, int, const char*, int, const char*, const string&, const char*, const char*, const char*, const char*, const char*, const char*)’
   34 | static void chisel_insertp(
      |             ^~~~~~~~~~~~~~
```

**Comment by oharboe** [comment]
Tried a different smaller test:

```
[info] ascenium.tiles.ALUTests *** ABORTED *** (1 second, 633 milliseconds)
[info]   java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter
[info]   at chiseltest.legacy.backends.verilator.VerilatorExecutive$.getCachedSim(VerilatorExecutive.scala:73)
[info]   at chiseltest.legacy.backends.verilator.VerilatorExecutive$.start(VerilatorExecutive.scala:200)
[info]   at chiseltest.defaults.package$.createDefaultTester(defaults.scala:24)
[info]   at chiseltest.ChiselScalatestTester$TestBuilder.apply(ChiselScalatestTester.scala:33)
[info]   at ascenium.tiles.ALUTests.$anonfun$new$5(ALUTests.scala:171)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   ...
[info]   Cause: java.lang.ClassNotFoundException: javax.xml.bind.DatatypeConverter
[info]   at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
[info]   at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
[info]   at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
[info]   at chiseltest.legacy.backends.verilator.VerilatorExecutive$.getCachedSim(VerilatorExecutive.scala:73)
[info]   at chiseltest.legacy.backends.verilator.VerilatorExecutive$.start(VerilatorExecutive.scala:200)
[info]   at chiseltest.defaults.package$.createDefaultTester(defaults.scala:24)
[info]   at chiseltest.ChiselScalatestTester$TestBuilder.apply(ChiselScalatestTester.scala:33)
[info]   at ascenium.tiles.ALUTests.$anonfun$new$5(ALUTests.scala:171)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
```

**Comment by oharboe** [comment]
J! First life! Was able to get caching for a small test. :-)

Some notes:

The first time I run the testsuite, I have many tests that use the same Verilator binary, I get no caching.

The second time, all tests are cached. 

My expectation is that during the first run of the testsuite, the Verilator binaries should be cached.

```
rm -rf test_run_dir
sbt "test:testOnly *ALUT*"
--- no Verilator binaries cached ---
sbt "test:testOnly *ALUT*"
--- all Verilator binaries cached ---
```

**Comment by oharboe** [comment]
I'm happy to set up an interactive Google Meet, if that's easier to figure out what's going on. I can't share our code easily...

**Comment by ekiwi** [comment]
> I finally ran into a problem that appears to be related to this new feature:

This is probably related to an earlier change. Could you show us the output of `verilator --version`?

**Comment by oharboe** [comment]
> > I finally ran into a problem that appears to be related to this new feature:
> 
> This is probably related to an earlier change. Could you show us the output of `verilator --version`?

This is the Ubuntu 20.04 verilator:

```
$ verilator --version
Verilator 4.028 2020-02-06 rev v4.026-92-g890cecc1
```

**Comment by ekiwi** [comment]
Awesome `4.0281 is still supported: https://github.com/ucb-bar/chisel-testers2/runs/2634410004?check_suite_focus=true

Does the project where you encounter this error have any special features that we might be missing in our CI? Like does it have black boxes, are you changing verilator flags? Anything out of the ordinary?

**Comment by oharboe** [comment]
> Does the project where you encounter this error have any special features that we might be missing in our CI? Like does it have black boxes, are you changing verilator flags? Anything out of the ordinary?

I tried to find out what adds the verilator `--coverage-user`  argument, but I can't find it...

Some extra environment variables I add, but no VerilatorFlags().

```
// faster tests by using ccache for Verilator
envVars in Test := Map(
  "OBJCACHE" -> "ccache",
  "MAKEFLAGS" -> "VM_PARALLEL_BUILDS=0 -j8"
)
```

**Comment by ekiwi** [comment]
> I tried to find out what adds the verilator `--coverage-user` argument, but I can't find it...

That is now done by default since we added the `cover` statement to firrtl and we want that to be supported out of the box.

**Comment by kammoh** [comment]
Use of `javax.xml.bind.DatatypeConverter` makes this incompatible with JRE versions > 8. For example on OpenJDK 11 I get this:
```
*** ABORTED ***
  java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter
  at chiseltest.legacy.backends.verilator.VerilatorExecutive$.getCachedSim(VerilatorExecutive.scala:73)
  at chiseltest.legacy.backends.verilator.VerilatorExecutive$.start(VerilatorExecutive.scala:200)
  at chiseltest.defaults.package$.createDefaultTester(defaults.scala:24)
  at chiseltest.ChiselScalatestTester$TestBuilder.apply(ChiselScalatestTester.scala:33)
```
Going through the code I'm not sure why conversion to hex string is even necessary? As far as I can see the hash is written and read in a file, so why not use a binary file?

**Comment by kammoh** [comment]
What do you about adding JDK 11 (at least) to the CI tests?

**Comment by kammoh** [comment]
Which tests check effectiveness of `CachingAnnotation`? #309 fixes the issue I had with Java 11, but caching doesn't seem to be working: `verilated` folder is apparently deleted each time `test` is run and everything is compiled from scratch. At the same time all tests are passing. Is it because this PR is not compatible with #303 or am I doing something wrong here?
